main/src/library/LazyList.flix

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2021 Jakob Schneider Villumsen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+pub enum LazyList[a] {
+    case ENil,
+    case ECons(a, LazyList[a]),
+    case LCons(a, Lazy[LazyList[a]]),
+    case LList(Lazy[LazyList[a]])
+}
+
+namespace LazyList {
+    // TODO: Implement functions used in TestExamples
+
+    ///
+    /// Returns an empty LazyList.
+    ///
+    pub def empty(): LazyList[a] = LList(lazy ENil)
+
+    ///
+    /// Returns true if and only if `l` is the empty LazyList, i.e. `ENil`.
+    ///
+    /// Does not force the tail of `l`.
+    ///
+    pub def isEmpty(l: LazyList[a]): Bool = match l {
+        case ENil      => true
+        case LList(ys) => isEmpty(force ys)
+        case _         => false
+    }
+
+    ///
+    /// Returns a lazy list of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns an empty lazy list if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): LazyList[Int32] =
+        if (b >= e)
+            empty()
+        else
+            LList(lazy LCons(b, lazy range(b + 1, e)))
+
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `l`.
+    ///
+    /// Returns `None` if `l` is empty.
+    ///
+    /// Does not force the tail of `l`.
+    ///
+    pub def head(l: LazyList[a]): Option[a] = match l {
+        case ENil        => None
+        case ECons(x, _) => Some(x)
+        case LCons(x, _) => Some(x)
+        case LList(ys)   => head(force ys)
+    }
+
+    ///
+    /// Returns the first `n` elements of `l`.
+    ///
+    /// Does not force the tail of `l`.
+    ///
+    pub def take(n: Int32, l: LazyList[a]): LazyList[a] =
+        if (n <= 0)
+            empty()
+        else
+            match l {
+                case ENil                     => ENil
+                case ECons(x, _)  if (n == 1) => LCons(x, lazy ENil)
```

Comment:
This case could just be omitted right?

Reply:
Suppose we have an infinite lazy list with natural numbers (from 0), `l`. The expression
`l |> LazyList.filter(x -> x == 3) |> LazyList.take(1) |> LazyList.foreach(x -> Console.printLine(Int32.toSring(x)))`
would then never terminate. The first element would be printed but the tail end would be stuck calling `filter` recursively forever.

Reply:
I don't follow: Would the recursive call not reduce `n` to `0` and thus complete?

Reply:
Followup: If you filter an infinite list, like in your example, I do agree it cannot terminate because you are asking for an element equal to `3` but no element every comes again that is equal to `3`.

Reply:
Yes you're completely right. I had it confused with a different structure. I'll tidy it up!
The old lazylist implementation also had an if-check with `n == 1`. Curios.

Reply:
Things get missed :)